### PR TITLE
Initial context

### PR DESCRIPTION
**Pull request by richardpringle**
- **Switch To Using Context**
- **gas**
- **update**
- **stash**
- **WIP**
- **Wrap `Program` in a `Context`**

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -37,6 +37,7 @@ require (
 require (
 	filippo.io/edwards25519 v1.0.0 // indirect
 	github.com/DataDog/zstd v1.5.2 // indirect
+	github.com/ava-labs/hypersdk/x/programs/cmd/simulator v0.0.0-20240131040904-4bd6d720a5bc // indirect
```
Comment (richardpringle) [inline go.mod:40]: 🤔

File: go.sum [hunk 2]
```diff
@@ -55,6 +55,8 @@ github.com/ava-labs/avalanche-network-runner v1.7.4-rc.0 h1:xNbCMNqenaDr0bb35j27
 github.com/ava-labs/avalanche-network-runner v1.7.4-rc.0/go.mod h1:B7Ynk/avkCk49CCIWbM4j1UrPlqIi0IHCPAB2MZNvLw=
 github.com/ava-labs/avalanchego v1.10.18 h1:ErJ+SJBtN9tVqk3OPRXffpf+MWeQnNZJlBNWQIgAg8A=
 github.com/ava-labs/avalanchego v1.10.18/go.mod h1:wqmokVcLXE+9Nfa8wHdGJtjEW35acVf8rkraSZPVBA4=
+github.com/ava-labs/hypersdk/x/programs/cmd/simulator v0.0.0-20240131040904-4bd6d720a5bc h1:JxdqCOsVLY4sD5Mh9R+zRc79GxRcXW0Cp7FB542O1KU=
+github.com/ava-labs/hypersdk/x/programs/cmd/simulator v0.0.0-20240131040904-4bd6d720a5bc/go.mod h1:cIFL1UbjI5LwYZCEjlThyiOWKW1Dh3T0tl8b+hbeV3M=
```
Comment (richardpringle) [inline go.sum:59]: 🤔

File: x/programs/cmd/simulator/vm/actions/program_execute.go [hunk 3]
```diff
@@ -107,13 +108,22 @@ func (t *ProgramExecute) Execute(
 	importsBuilder.Register("state", func() host.Import {
 		return pstate.New(logging.NoLog{}, mu)
 	})
+	callContext := program.Context{
+		ProgramID: programID,
+		// Actor:            [32]byte(actor[1:]),
+		// OriginatingActor: [32]byte(actor[1:])
```
Comment (richardpringle) [inline x/programs/cmd/simulator/vm/actions/program_execute.go:114]: The change is big enough already that I want to add this part in a new PR

File: x/programs/cmd/simulator/vm/actions/program_execute.go [hunk 4]
```diff
@@ -107,13 +108,22 @@ func (t *ProgramExecute) Execute(
...
+	}
+
+	// print callContext to stderr
+	fmt.Fprintf(os.Stderr, "\ncallContext: %+v\n", callContext)
+
```
Comment (richardpringle) [inline x/programs/cmd/simulator/vm/actions/program_execute.go:119]:
```suggestion
```

File: x/programs/examples/imports/program/program.go [hunk 5]
```diff
@@ -23,171 +23,171 @@ import (
...
 	meter   *engine.Meter
 	imports host.SupportedImports
+	ctx     *program.Context
 }
 
 // New returns a new program invoke host module which can perform program to program calls.
-func New(log logging.Logger, engine *engine.Engine, mu state.Mutable, cfg *runtime.Config) *Import {
+func New(log logging.Logger, engine *engine.Engine, mu state.Mutable, cfg *runtime.Config, ctx *program.Context) *Import {
 	return &Import{
+		name:   "program", // TODO fix this
```
Comment (richardpringle) [inline x/programs/examples/imports/program/program.go:41]: delete this

File: x/programs/examples/imports/program/program.go [hunk 6]
```diff
@@ -23,171 +23,171 @@ import (
...
+func (i *Import) Name() string {
+	return i.name
 }
 
-func (i *Import) Register(link *host.Link) error {
+func (i *Import) Register(link *host.Link, callContext program.Context) error {
 	i.meter = link.Meter()
 	i.imports = link.Imports()
-	return link.RegisterImportFn(Name, "call_program", i.callProgramFn)
+	return link.RegisterImportFn(i.name, "call_program", i.callProgramFn(callContext))
```
Comment (richardpringle) [inline x/programs/examples/imports/program/program.go:57]: Put this back

File: x/programs/examples/token.go [hunk 7]
```diff
@@ -33,8 +34,9 @@ const (
 	Balance
 )
 
-func NewToken(log logging.Logger, engine *engine.Engine, programBytes []byte, db state.Mutable, cfg *runtime.Config, imports host.SupportedImports, maxUnits uint64) *Token {
+func NewToken(id ids.ID, log logging.Logger, engine *engine.Engine, programBytes []byte, db state.Mutable, cfg *runtime.Config, imports host.SupportedImports, maxUnits uint64) *Token {
 	return &Token{
+		id:           id,
```
Comment (richardpringle) [inline x/programs/examples/token.go:39]: Not sure if this makes sense. 

But I also don't quite understand why there is so much Go code to define "Token" like this. This should all only be in Rust IMO

File: x/programs/examples/imports/pstate/pstate.go [hunk 8]
```diff
@@ -40,7 +40,7 @@ func (*Import) Name() string {
 	return Name
 }
 
-func (i *Import) Register(link *host.Link) error {
+func (i *Import) Register(link *host.Link, _ program.Context) error {
```
Comment (richardpringle) [inline x/programs/examples/imports/pstate/pstate.go:43]: 🤔 I'm not sure why/how this was going to be used.

File: x/programs/examples/imports/program/program.go [hunk 9]
```diff
@@ -202,6 +202,7 @@ func getCallArgs(_ context.Context, memory *program.Memory, buffer []byte, progr
 		if err != nil {
 			return nil, err
 		}
+		// argPtr, err := program.NewSmartPtr(uint32(ptr), int(length))
```
Comment (richardpringle) [inline x/programs/examples/imports/program/program.go:205]: I think I can delete this

File: x/programs/examples/counter_test.go [hunk 10]
```diff
@@ -78,58 +82,58 @@ func TestCounterProgram(t *testing.T) {
...
+	// programID2Ptr, err := argumentToSmartPtr(programID2, mem2)
 	require.NoError(err)
 
 	// write alice's key to stack and get pointer
 	alicePtr2, err := argumentToSmartPtr(alicePublicKey, mem2)
 	require.NoError(err)
 
 	// initialize counter for alice on runtime 2
-	result, err = rt2.Call(ctx, "initialize_address", programID2Ptr, alicePtr2)
+	result, err = rt2.Call(ctx, "initialize_address", program.Context{ProgramID: programID2}, alicePtr2)
```
Comment (richardpringle) [inline x/programs/examples/counter_test.go:124]: Shouldn't need to recreate this `Context` every single time.

File: x/programs/examples/utils.go [hunk 11]
```diff
@@ -24,15 +24,15 @@ func newKey() (ed25519.PublicKey, error) {
...
-// SerializeParameter serializes [obj] using Borsh
-func serializeParameter(obj interface{}) ([]byte, error) {
-	bytes, err := borsh.Serialize(obj)
-	return bytes, err
-}
+// // SerializeParameter serializes [obj] using Borsh
+// func serializeParameter(obj interface{}) ([]byte, error) {
+// 	bytes, err := borsh.Serialize(obj)
+// 	return bytes, err
+// }
```
Comment (richardpringle) [inline x/programs/examples/utils.go:31]: Delete this

File: x/programs/host/dependencies.go [hunk 12]
```diff
@@ -3,9 +3,11 @@
...
 package host
 
+import "github.com/ava-labs/hypersdk/x/programs/program"
+
 type Import interface {
 	// Name returns the name of this import module.
 	Name() string
 	// Register registers this import module with the provided link.
-	Register(*Link) error
+	Register(*Link, program.Context) error
```
Comment (richardpringle) [inline x/programs/host/dependencies.go:12]: This isn't used (yet), can put this back

File: x/programs/host/link.go [hunk 13]
```diff
@@ -58,7 +58,7 @@ func (l *Link) Instantiate(store *engine.Store, mod *wasmtime.Module, cb ImportF
 		if !ok {
 			return nil, fmt.Errorf("%w: %s", ErrMissingImportModule, imp)
 		}
-		err := importFn().Register(l)
+		err := importFn().Register(l, callContext)
```
Comment (richardpringle) [inline x/programs/host/link.go:61]: Same here?

File: Cargo.toml
Comment (richardpringle) [inline Cargo.toml:1]: Add test crate to test that the macros are behaving as expected and that the code can be compiled to wasm and executed.

File: x/programs/cmd/simulator/cmd/plan.go
Comment (richardpringle) [inline x/programs/cmd/simulator/cmd/plan.go:1]: I think a formatter/linter did this

File: x/programs/cmd/simulator/cmd/program.go
Comment (richardpringle) [inline x/programs/cmd/simulator/cmd/program.go:1]: Added things, then deleted them. The empty lines stayed.

File: x/programs/cmd/simulator/vm/storage/state_manager.go
Comment (richardpringle) [inline x/programs/cmd/simulator/vm/storage/state_manager.go:1]: formatter

File: x/programs/examples/imports/program/program.go
Comment (richardpringle) [inline x/programs/examples/imports/program/program.go:1]: The major change here is that the `CallProgramFn` is now a method that takes the _context_ as input and returns a closure to call the wasm-function with the given _context_. Other changes are minimal.

File: x/programs/examples/token.go
Comment (richardpringle) [inline x/programs/examples/token.go:1]: Adhere to the new APIs. I'm not quite sure what the purpose of this file is (similar to  my comment on `counter_test.go`).

File: x/programs/examples/token_test.go
Comment (richardpringle) [inline x/programs/examples/token_test.go:1]: Adhere to new APIs. Same comment as the other two files, I don't quite understand the benefit of this file. I do know that some of the delete functionality is being tested in here (unchanged in this PR), but there should be a better way to test that.

File: x/programs/host/link_test.go
Comment (richardpringle) [inline x/programs/host/link_test.go:1]: Adhere to the new APIs

File: x/programs/rust/examples/counter/src/lib.rs
Comment (richardpringle) [inline x/programs/rust/examples/counter/src/lib.rs:1]: Use `Context` struct instead of `Program`

File: x/programs/rust/examples/token/src/lib.rs
Comment (richardpringle) [inline x/programs/rust/examples/token/src/lib.rs:1]: Use `Context` struct instead of `Program`

File: x/programs/examples/token.go [hunk 14]
```diff
@@ -67,12 +67,12 @@ type Token struct {
 
 func (t *Token) Context() program.Context {
 	return program.Context{
-		ProgramID: t.id,
+		ProgramID: t.programID,
 	}
 }
 
 func (t *Token) ID() ids.ID {
```
Comment (darioush) [inline x/programs/examples/token.go:74]: naming nit: ProgramID()?

File: x/programs/rust/wasmlanche-sdk/tests/test-crate/src/lib.rs [hunk 15]
```diff
@@ -3,17 +3,16 @@
...
 
 #[public]
-pub fn noop(_: Context) -> i64 {
+pub fn always_true(_: Context) -> i64 {
     true as i64
 }
 
 #[public]
-pub fn return_id(context: Context) -> u32 {
+pub fn combine_last_bit_of_each_id_byte(context: Context) -> u32 {
```
Comment (darioush) [inline x/programs/rust/wasmlanche-sdk/tests/test-crate/src/lib.rs:11]: naming nit: nicer, perhaps `calculate_id_checksum`?

File: x/programs/rust/wasmlanche-sdk/tests/public_function.rs [hunk 16]
```diff
@@ -1,98 +1,132 @@
...
+        let context = Context { program };
+        let serialized_context = borsh::to_vec(&context).expect("failed to serialize context");
+
+        test_crate.allocate(serialized_context)
+    };
+
+    assert!(test_crate.always_true(context_ptr));
+
+    let combined_binary_digits = test_crate.combine_last_bit_of_each_id_byte(context_ptr);
+    assert_eq!(combined_binary_digits, u32::MAX);
```
Comment (darioush) [inline x/programs/rust/wasmlanche-sdk/tests/public_function.rs:63]: why do we expect this to be true?

File: x/programs/examples/imports/program/program.go [hunk 17]
```diff
@@ -33,161 +33,161 @@ type Import struct {
 	engine  *engine.Engine
 	meter   *engine.Meter
 	imports host.SupportedImports
+	ctx     *program.Context
 }
 
 // New returns a new program invoke host module which can perform program to program calls.
-func New(log logging.Logger, engine *engine.Engine, mu state.Mutable, cfg *runtime.Config) *Import {
+func New(log logging.Logger, engine *engine.Engine, mu state.Mutable, cfg *runtime.Config, ctx *program.Context) *Import {
```
Comment (darioush) [inline x/programs/examples/imports/program/program.go:40]: nit: should we also take `program.Context` (not ptr) here? (overall seems switching to ptr would be better, your preference if you want to make a separate change)

File: x/programs/program/dependencies.go [hunk 18]
```diff
@@ -3,11 +3,14 @@
...
+import "github.com/bytecodealliance/wasmtime-go/v14"
+
 type Instance interface {
 	// GetFunc returns a function exported by the program.
 	GetFunc(name string) (*Func, error)
 	// GetExport returns an export exported by the program.
 	GetExport(name string) (*Export, error)
 	// Memory returns the memory exported by the program.
 	Memory() (*Memory, error)
+	GetStore() wasmtime.Storelike
```
Comment (darioush) [inline x/programs/program/dependencies.go:15]: nit could we add a comment (what is Store?)

File: x/programs/runtime/runtime.go [hunk 19]
```diff
@@ -45,7 +45,7 @@ type WasmRuntime struct {
 	log logging.Logger
 }
 
-func (r *WasmRuntime) Initialize(ctx context.Context, programBytes []byte, maxUnits uint64) (err error) {
+func (r *WasmRuntime) Initialize(ctx context.Context, callContext program.Context, programBytes []byte, maxUnits uint64) (err error) {
```
Comment (darioush) [inline x/programs/runtime/runtime.go:48]: nit: this is my (subjective) preference for the argument order, (passing any ctx like object either first or right after golang's `context.Context`), but overall it seems we can be more consistent

File: x/programs/examples/imports/program/program.go [hunk 20]
```diff
@@ -33,161 +33,161 @@ type Import struct {
...
+				i.log.Error("failed to transfer remaining balance to caller",
+					zap.Error(err),
+				)
+			}
+		}()
+
+		argsBytes, err := program.SmartPtr(args).Bytes(memory)
 		if err != nil {
-			i.log.Error("failed to transfer remaining balance to caller",
+			i.log.Error("failed to read program args name from memory",
```
Comment (darioush) [inline x/programs/examples/imports/program/program.go:147]:
```suggestion
			i.log.Error("failed to read program args from memory",
```
?

File: x/programs/examples/imports/program/program.go [hunk 21]
```diff
@@ -33,161 +33,161 @@ type Import struct {
...
+			return -1
+		}
 
-	res, err := rt.Call(ctx, string(functionBytes), params...)
-	if err != nil {
-		i.log.Error("failed to call entry function",
-			zap.Error(err),
-		)
-		return -1
+		// return int64(res[0])
```
Comment (darioush) [inline x/programs/examples/imports/program/program.go:183]:
```suggestion
```

**Review by darioush — APPROVED** [review]

File: x/programs/program/dependencies.go (hunk 18 above)
Comment (richardpringle) [reply x/programs/program/dependencies.go:15]: This was one of @dboehm-avalabs's changes. I'm going to leave it for now.

File: x/programs/examples/imports/program/program.go (hunk 17 above)
Comment (richardpringle) [reply x/programs/examples/imports/program/program.go:40]: Going to leave this here and pass the pointer everywhere else (in a follow up)

File: x/programs/rust/wasmlanche-sdk/tests/test-crate/src/lib.rs (hunk 15 above)
Comment (richardpringle) [reply x/programs/rust/wasmlanche-sdk/tests/test-crate/src/lib.rs:11]: Calling this a checksum is a bit of a stretch haha. But that is more or less what I'm trying to do. This test isn't great though, I'll try and think of a better way to test this.

File: x/programs/examples/imports/pstate/pstate.go (hunk 8 above)
Comment (wlawt) [reply x/programs/examples/imports/pstate/pstate.go:43]: I might've missed something, but where's the other `Register` implementation?

**Review by darioush — APPROVED** [review]

**Pull request merged**